File: test/fake_prometheus.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

/// A stand-in Prometheus built from real responses captured off the live server
/// (Proxmox host + 4 guests, a Tesla V100 whose dcgm-exporter is down).
///
/// [gpuUp] flips the GPU exporter between live and long-dead so both branches of
/// the GPU panel can be rendered in tests.
class FakePrometheus {
  FakePrometheus({this.gpuUp = false, this.failEverything = false});

  final bool gpuUp;
  final bool failEverything;

  int instantCalls = 0;
  int rangeCalls = 0;
  final List<String> queries = [];

  static const _nodes = <String, String>{
    'pve-host': 'hypervisor',
    'vm-node-1': 'nodes',
    'vm-ops-node': 'operations',
    'vm-vpn': 'vpn',
    'vm-amnezia-proxy': 'proxy',
  };

  static const _cpu = {
    'pve-host': 9.19,
    'vm-node-1': 10.60,
    'vm-ops-node': 0.67,
    'vm-vpn': 0.33,
    'vm-amnezia-proxy': 0.38,
  };
  static const _cores = {
    'pve-host': 12.0,
    'vm-node-1': 2.0,
    'vm-ops-node': 2.0,
    'vm-vpn': 1.0,
    'vm-amnezia-proxy': 1.0,
  };
  static const _memTotal = {
    'pve-host': 33572134912.0,
    'vm-node-1': 8257101824.0,
    'vm-ops-node': 2978578432.0,
    'vm-vpn': 937627648.0,
    'vm-amnezia-proxy': 937627648.0,
  };
  static const _memAvail = {
    'pve-host': 17745612800.0,
    'vm-node-1': 2990116864.0,
    'vm-ops-node': 2232578048.0,
    'vm-vpn': 517435392.0,
    'vm-amnezia-proxy': 509181952.0,
  };
  static const _fsSize = {
    'pve-host': 100861726720.0,
    'vm-node-1': 65445814272.0,
    'vm-ops-node': 65445814272.0,
    'vm-vpn': 9283444736.0,
    'vm-amnezia-proxy': 15523123200.0,
  };
  static const _fsAvail = {
    'pve-host': 63221596160.0,
    'vm-node-1': 51153305600.0,
    'vm-ops-node': 53294276608.0,
    'vm-vpn': 5926871040.0,
    'vm-amnezia-proxy': 12062707712.0,
  };

  http.Client client() => MockClient((req) async {
    if (failEverything) {
      return http.Response('connection refused', 503);
    }
    final q = req.url.queryParameters['query'] ?? '';
    queries.add(q);
    if (req.url.path.endsWith('query_range')) {
      rangeCalls++;
      return _json(_rangeFor(q));
    }
    instantCalls++;
    return _json({'resultType': 'vector', 'result': _instantFor(q)});
  });

  http.Response _json(Map<String, dynamic> data) => http.Response(
    jsonEncode({'status': 'success', 'data': data}),
    200,
    headers: {'content-type': 'application/json'},
  );

  double get _now => DateTime.now().millisecondsSinceEpoch / 1000;

  List<Map<String, dynamic>> _vector(
    Map<String, double> byInstance, {
    String name = 'x',
    Map<String, String> extra = const {},
  }) => [
    for (final e in byInstance.entries)
      {
        'metric': {
          '__name__': name,
          'instance': e.key,
          'job': 'node',
          'role': _nodes[e.key] ?? '-',
          ...extra,
        },
        'value': [_now, e.value.toString()],
      },
  ];

  List<Map<String, dynamic>> _gpuVector(String name, double? value) {
    if (value == null) return const [];
    return [
      {
        'metric': {
          '__name__': name,
          'instance': 'vm-gpu-worker-1',
          'job': 'dcgm',
          'role': 'gpu',
          'gpu': '0',
          'device': 'nvidia0',
          'modelName': 'Tesla V100-SXM2-16GB',
          'UUID': 'GPU-0f5424e3-7915-abf0-4445-c5206f1ed148',
        },
        'value': [_now, value.toString()],
      },
    ];
  }

  List<Map<String, dynamic>> _instantFor(String q) {
    if (q == 'up') {
      return [
        ..._vector({for (final k in _nodes.keys) k: 1}, name: 'up'),
        {
          'metric': {
            '__name__': 'up',
            'instance': 'vm-gpu-worker-1',
            'job': 'dcgm',
            'role': 'gpu',
          },
          'value': [_now, gpuUp ? '1' : '0'],
        },
        {
          'metric': {
            '__name__': 'up',
            'instance': 'localhost:9090',
            'job': 'prometheus',
          },
          'value': [_now, '1'],
        },
      ];
    }
    if (q.contains('node_hwmon_sensor_label')) {
      return [
        _temp('Tctl', 'temp1', 43.375),
        _temp('Tccd1', 'temp3', 46.5),
      ];
    }
    if (q.startsWith('100 - ')) return _vector(_cpu);
    if (q.contains('mode="iowait"')) {
      return _vector({for (final k in _nodes.keys) k: 0.12});
    }
    if (q.startsWith('count by')) return _vector(_cores);
    if (q == 'node_memory_MemTotal_bytes') return _vector(_memTotal);
    if (q == 'node_memory_MemAvailable_bytes') return _vector(_memAvail);
    if (q == 'node_memory_SwapTotal_bytes') {
      return _vector({for (final k in _nodes.keys) k: k == 'pve-host' ? 8589930496.0 : 0.0});
    }
    if (q == 'node_memory_SwapFree_bytes') {
      return _vector({for (final k in _nodes.keys) k: k == 'pve-host' ? 8589930496.0 : 0.0});
    }
    if (q == 'node_load1') return _vector({for (final k in _nodes.keys) k: 0.66});
    if (q == 'node_load5') return _vector({for (final k in _nodes.keys) k: 0.71});
    if (q == 'node_load15') return _vector({for (final k in _nodes.keys) k: 0.60});
    if (q == 'node_boot_time_seconds') {
      return _vector({for (final k in _nodes.keys) k: _now - 180000});
    }
    if (q.startsWith('node_filesystem_size')) return _vector(_fsSize);
    if (q.startsWith('node_filesystem_avail')) return _vector(_fsAvail);
    if (q.contains('receive_bytes')) {
      return _vector({for (final k in _nodes.keys) k: 5574.87});
    }
    if (q.contains('transmit_bytes')) {
      return _vector({for (final k in _nodes.keys) k: 1103.31});
    }

    // GPU. Values come back through last_over_time even while the target is
    // down, which is exactly the case the UI has to handle.
    if (q.contains('DCGM_FI_DEV_GPU_TEMP')) {
      if (q.startsWith('time()')) {
        return _gpuVector('gpu_age', gpuUp ? 12.0 : 166055.5);
      }
      return _gpuVector('DCGM_FI_DEV_GPU_TEMP', 40);
    }
    if (q.contains('DCGM_FI_DEV_GPU_UTIL')) {
      return _gpuVector('DCGM_FI_DEV_GPU_UTIL', gpuUp ? 73 : 0);
    }
    if (q.contains('DCGM_FI_DEV_MEMORY_TEMP')) {
      return _gpuVector('DCGM_FI_DEV_MEMORY_TEMP', 38);
    }
    if (q.contains('DCGM_FI_DEV_FB_USED')) {
      return _gpuVector('DCGM_FI_DEV_FB_USED', gpuUp ? 9216 : 0);
    }
    if (q.contains('DCGM_FI_DEV_FB_FREE')) {
      return _gpuVector('DCGM_FI_DEV_FB_FREE', gpuUp ? 7154 : 16370);
    }
    if (q.contains('DCGM_FI_DEV_POWER_USAGE')) {
      return _gpuVector('DCGM_FI_DEV_POWER_USAGE', gpuUp ? 212.4 : 24.9);
    }
    if (q.contains('DCGM_FI_DEV_SM_CLOCK')) {
      return _gpuVector('DCGM_FI_DEV_SM_CLOCK', gpuUp ? 1380 : 135);
    }
    if (q.contains('DCGM_FI_DEV_MEM_CLOCK')) {
      return _gpuVector('DCGM_FI_DEV_MEM_CLOCK', 877);
    }
    if (q.contains('DCGM_FI_DEV_MEM_COPY_UTIL')) {
      return _gpuVector('DCGM_FI_DEV_MEM_COPY_UTIL', gpuUp ? 41 : 0);
    }
    return const [];
  }

  Map<String, dynamic> _rangeFor(String q) {
    final end = _now;
    List<List<Object>> pts(double base) => [
      for (var i = 60; i >= 0; i--)
        [end - i * 60, (base + (i % 7) - 3).toStringAsFixed(2)],
    ];
    if (q.contains('DCGM')) {
      return {
        'resultType': 'matrix',
        'result': [
          {
            'metric': {
              'instance': 'vm-gpu-worker-1',
              'gpu': '0',
              'modelName': 'Tesla V100-SXM2-16GB',
            },
            'values': pts(65),
          },
        ],
      };
    }
    if (q.contains('node_hwmon')) {
      return {
        'resultType': 'matrix',
        'result': [
          {
            'metric': {'instance': 'pve-host', 'label': 'Tctl', 'sensor': 'temp1'},
            'values': pts(44),
          },
        ],
      };
    }
    return {
      'resultType': 'matrix',
      'result': [
        for (final e in _nodes.entries)
          {
            'metric': {'instance': e.key, 'role': e.value},
            'values': pts(_cpu[e.key]! + 10),
          },
      ],
    };
  }

  Map<String, dynamic> _temp(String label, String sensor, double c) => {
    'metric': {
      '__name__': 'node_hwmon_temp_celsius',
      'instance': 'pve-host',
      'job': 'node',
      'role': 'hypervisor',
      'chip': 'pci0000:00_0000:00:18_3',
      'sensor': sensor,
      'label': label,
    },
    'value': [_now, c.toString()],
  };
}

File: tests/fake_prometheus.rs
//! Local stand-in for the Proxmox host, four guests and a V100 worker.
use chrono::Utc;
use rackglass::{Config, PromClient, prom::queries as q};
use serde_json::{Value, json};
use std::{
    collections::HashMap,
    io::{Read, Write},
    net::{TcpListener, TcpStream},
    sync::{
        Arc, Mutex,
        atomic::{AtomicBool, Ordering},
    },
    thread::{self, JoinHandle},
    time::Duration,
};

#[derive(Clone, Debug)]
pub struct Reply {
    pub status: u16,
    pub body: String,
}
impl Reply {
    pub fn vector(rows: Vec<Value>) -> Self {
        Self {
            status: 200,
            body: json!({"status":"success","data":{"resultType":"vector","result":rows}})
                .to_string(),
        }
    }
}
#[derive(Default)]
pub struct Data {
    pub replies: HashMap<String, Reply>,
    pub requests: Vec<(String, HashMap<String, String>)>,
    pub fail: bool,
    pub delay: Duration,
}
pub struct FakePrometheus {
    pub url: String,
    pub data: Arc<Mutex<Data>>,
    stop: Arc<AtomicBool>,
    worker: Option<JoinHandle<()>>,
}
impl FakePrometheus {
    pub fn new(gpu_up: bool) -> Self {
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        let url = format!("http://{}", listener.local_addr().unwrap());
        let data = Arc::new(Mutex::new(Data {
            replies: cluster(gpu_up),
            ..Default::default()
        }));
        let stop = Arc::new(AtomicBool::new(false));
        let d = data.clone();
        let s = stop.clone();
        let worker = thread::spawn(move || {
            let mut workers = Vec::new();
            for stream in listener.incoming() {
                if s.load(Ordering::Acquire) {
                    break;
                }
                let d = d.clone();
                workers.push(thread::spawn(move || serve(stream.unwrap(), &d)));
            }
            for w in workers {
                w.join().unwrap();
            }
        });
        Self {
            url,
            data,
            stop,
            worker: Some(worker),
        }
    }
    pub fn client(&self) -> PromClient {
        PromClient::new(&self.url)
    }
    pub fn set(&self, query: impl Into<String>, rows: Vec<Value>) {
        self.data
            .lock()
            .unwrap()
            .replies
            .insert(query.into(), Reply::vector(rows));
    }
    pub fn reply(&self, query: &str, status: u16, body: &str) {
        self.data.lock().unwrap().replies.insert(
            query.into(),
            Reply {
                status,
                body: body.into(),
            },
        );
    }
    pub fn calls(&self) -> usize {
        self.data.lock().unwrap().requests.len()
    }
    pub fn queries(&self) -> Vec<String> {
        self.data
            .lock()
            .unwrap()
            .requests
            .iter()
            .map(|(_, p)| p["query"].clone())
            .collect()
    }
}
impl Drop for FakePrometheus {
    fn drop(&mut self) {
        self.stop.store(true, Ordering::Release);
        let _ = TcpStream::connect(self.url.trim_start_matches("http://"));
        if let Some(w) = self.worker.take() {
            w.join().unwrap();
        }
    }
}
fn decode(s: &str) -> String {
    let mut bytes = Vec::new();
    let mut it = s.bytes();
    while let Some(b) = it.next() {
        if b == b'%' {
            let a = it.next().unwrap();
            let b = it.next().unwrap();
            bytes.push(u8::from_str_radix(std::str::from_utf8(&[a, b]).unwrap(), 16).unwrap());
        } else {
            bytes.push(if b == b'+' { b' ' } else { b });
        }
    }
    String::from_utf8(bytes).unwrap()
}
fn serve(mut stream: TcpStream, data: &Mutex<Data>) {
    stream
        .set_read_timeout(Some(Duration::from_secs(5)))
        .unwrap();
    let mut request = Vec::new();
    let mut buf = [0; 1024];
    while !request.windows(4).any(|w| w == b"\r\n\r\n") {
        let n = stream.read(&mut buf).unwrap();
        if n == 0 {
            return;
        }
        request.extend_from_slice(&buf[..n]);
    }
    let text = String::from_utf8(request).unwrap();
    let target = text.split_whitespace().nth(1).unwrap();
    let (path, qs) = target.split_once('?').unwrap();
    let params: HashMap<_, _> = qs
        .split('&')
        .map(|p| {
            let (k, v) = p.split_once('=').unwrap();
            (decode(k), decode(v))
        })
        .collect();
    let (reply, delay) = {
        let mut d = data.lock().unwrap();
        d.requests.push((path.into(), params.clone()));
        let reply = if d.fail {
            Reply {
                status: 503,
                body: "connection refused".into(),
            }
        } else if path.ends_with("query_range") {
            range_reply(&params["query"], params["end"].parse().unwrap())
        } else {
            d.replies
                .get(&params["query"])
                .cloned()
                .unwrap_or_else(|| Reply::vector(vec![]))
        };
        (reply, d.delay)
    };
    thread::sleep(delay);
    let _ = write!(
        stream,
        "HTTP/1.1 {} Response\r\nContent-Type: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
        reply.status,
        reply.body.len(),
        reply.body
    );
}
pub fn sample(labels: Value, v: f64) -> Value {
    json!({"metric":labels,"value":[Utc::now().timestamp_millis() as f64/1000.0,v.to_string()]})
}
pub fn gpu(v: f64) -> Value {
    sample(
        json!({"instance":"vm-gpu-worker-1","job":"dcgm","gpu":"0","device":"nvidia0","modelName":"Tesla V100-SXM2-16GB","UUID":"GPU-v100"}),
        v,
    )
}
pub fn up(gpu_up: bool) -> Vec<Value> {
    let mut rows: Vec<_> = NODES
        .iter()
        .map(|(i, r)| sample(json!({"instance":i,"role":r,"job":"node"}), 1.0))
        .collect();
    rows.push(sample(
        json!({"instance":"vm-gpu-worker-1","role":"gpu-workers","job":"node"}),
        if gpu_up { 1.0 } else { 0.0 },
    ));
    rows.push(sample(
        json!({"instance":"vm-gpu-worker-1","role":"gpu","job":"dcgm"}),
        if gpu_up { 1.0 } else { 0.0 },
    ));
    rows.push(sample(
        json!({"instance":"localhost:9090","job":"prometheus"}),
        1.0,
    ));
    rows
}
const NODES: [(&str, &str); 5] = [
    ("pve-host", "hypervisor"),
    ("vm-node-1", "nodes"),
    ("vm-ops-node", "operations"),
    ("vm-vpn", "vpn"),
    ("vm-amnezia-proxy", "proxy"),
];
fn vector(values: [f64; 5]) -> Vec<Value> {
    NODES
        .iter()
        .zip(values)
        .map(|((i, r), v)| sample(json!({"instance":i,"role":r,"job":"node"}), v))
        .collect()
}
fn cluster(gpu_up: bool) -> HashMap<String, Reply> {
    let cfg = Config::default();
    let mut m = HashMap::new();
    for (_, q) in q::instant_poll_queries(&cfg)
        .into_iter()
        .chain(q::gpu_fallback_queries())
    {
        m.insert(q, Reply::vector(vec![]));
    }
    let mut set = |q: &str, rows| {
        m.insert(q.into(), Reply::vector(rows));
    };
    set(q::UP, up(gpu_up));
    for (query, values) in [
        (q::CPU_BUSY, [9.19, 10.60, 0.67, 0.33, 0.38]),
        (q::CORES, [12.0, 2.0, 2.0, 1.0, 1.0]),
        (
            q::MEM_TOTAL,
            [
                33572134912.0,
                8257101824.0,
                2978578432.0,
                937627648.0,
                937627648.0,
            ],
        ),
        (
            q::MEM_AVAILABLE,
            [
                17745612800.0,
                2990116864.0,
                2232578048.0,
                517435392.0,
                509181952.0,
            ],
        ),
        (
            q::FS_SIZE,
            [
                100861726720.0,
                65445814272.0,
                65445814272.0,
                9283444736.0,
                15523123200.0,
            ],
        ),
        (
            q::FS_AVAIL,
            [
                63221596160.0,
                51153305600.0,
                53294276608.0,
                5926871040.0,
                12062707712.0,
            ],
        ),
        (q::SWAP_TOTAL, [8589930496.0, 0.0, 0.0, 0.0, 0.0]),
        (q::SWAP_FREE, [8589930496.0, 0.0, 0.0, 0.0, 0.0]),
        (q::LOAD1, [0.66; 5]),
        (q::LOAD5, [0.71; 5]),
        (q::LOAD15, [0.60; 5]),
        (q::CPU_IO_WAIT, [0.12; 5]),
        (q::BOOT_TIME, [Utc::now().timestamp() as f64 - 180000.0; 5]),
    ] {
        set(query, vector(values));
    }
    set(&q::net_rx(&cfg), vector([5574.87; 5]));
    set(&q::net_tx(&cfg), vector([1103.31; 5]));
    set(
        q::ALL_TEMPS,
        vec![
            sample(
                json!({"instance":"pve-host","chip":"pci0000:00_0000:00:18_3","sensor":"temp3","label":"Tccd1"}),
                46.5,
            ),
            sample(
                json!({"instance":"pve-host","chip":"pci0000:00_0000:00:18_3","sensor":"temp7"}),
                39.25,
            ),
            sample(
                json!({"instance":"pve-host","chip":"pci0000:00_0000:00:18_3","sensor":"temp1","label":"Tctl"}),
                43.375,
            ),
        ],
    );
    for (metric, live, dead) in [
        ("DCGM_FI_DEV_GPU_TEMP", 40.0, 40.0),
        ("DCGM_FI_DEV_GPU_UTIL", 73.0, 0.0),
        ("DCGM_FI_DEV_MEMORY_TEMP", 38.0, 38.0),
        ("DCGM_FI_DEV_FB_USED", 9216.0, 0.0),
        ("DCGM_FI_DEV_FB_FREE", 7154.0, 16370.0),
        ("DCGM_FI_DEV_POWER_USAGE", 212.4, 24.9),
        ("DCGM_FI_DEV_SM_CLOCK", 1380.0, 135.0),
        ("DCGM_FI_DEV_MEM_CLOCK", 877.0, 877.0),
    ] {
        set(
            &q::fresh_gpu(metric),
            vec![gpu(if gpu_up { live } else { dead })],
        );
        set(&q::last_gpu(metric), vec![gpu(dead)]);
    }
    set(
        q::GPU_AGE_FRESH,
        vec![gpu(if gpu_up { 12.0 } else { 166055.5 })],
    );
    set(q::GPU_AGE_DEEP, vec![gpu(166055.5)]);
    m
}

fn range_reply(query: &str, end: f64) -> Reply {
    let points = |base: f64| -> Vec<Value> {
        (0..=240)
            .rev()
            .map(|i| {
                json!([
                    end - i as f64 * 15.0,
                    format!("{:.2}", base + (i % 7) as f64 - 3.0)
                ])
            })
            .collect()
    };
    let rows: Vec<Value> = if query.contains("DCGM") {
        vec![
            json!({"metric":{"instance":"vm-gpu-worker-1","gpu":"0","modelName":"Tesla V100-SXM2-16GB"},"values":points(65.0)}),
        ]
    } else if query.contains("speedtest_download") || query.contains("speedtest_upload") {
        let bases = if query.contains("upload") {
            [431e6, 66e6]
        } else {
            [222e6, 165e6]
        };
        ["direct","socks"].into_iter().zip(bases).map(|(path,base)| {
            let values: Vec<_> = (0..=240).rev().map(|i| json!([end-i as f64*15.0,format!("{:.0}",base+(i%5) as f64*1e6)])).collect();
            json!({"metric":{"instance":"vm-ops-node","job":"node","path":path},"values":values})
        }).collect()
    } else if query.contains("node_hwmon") {
        vec![
            json!({"metric":{"instance":"pve-host","label":"Tctl","sensor":"temp1"},"values":points(44.0)}),
        ]
    } else {
        let pinned = query
            .split_once("instance=\"")
            .and_then(|(_, s)| s.split_once('"').map(|(i, _)| i));
        NODES
            .iter()
            .zip([9.19, 10.60, 0.67, 0.33, 0.38])
            .filter(|((i, _), _)| pinned.is_none_or(|p| p == *i))
            .map(
                |((i, r), cpu)| json!({"metric":{"instance":i,"role":r},"values":points(cpu+10.0)}),
            )
            .collect()
    };
    Reply {
        status: 200,
        body: json!({"status":"success","data":{"resultType":"matrix","result":rows}}).to_string(),
    }
}
